02-intermediate-ruby/csv-as-a-database.md

```diff
@@ -14,136 +57,292 @@ A _database_ is a system of storing and organizing information on a computer. In
-  csv << [5, "Jupiter", 318, 5.2]
-end
+
+planets = CSV.read('planets_data.csv', headers: true)
+
+ap planets
+# => #<CSV::Table mode:col_or_row row_count:5>
 ```
 
-We can iterate over each line of the file using the `each` method:
+This difference will be useful in the next step...
+
+### Chaining `.each` Lets Us Look At One Record at a Time
+
+We can chain `.each` to `CSV.read` and a whole `each` code block at the end of this line.
+
+Observe this code:
 
 ```ruby
+require 'awesome_print'
 require 'csv'
-CSV.open("planet_data.csv", 'r').each do |line|
-  puts line
+
+CSV.read('planets_data.csv', headers: true).each do |record|
+  ap "Hooray!"
+  ap record
 end
 ```
 
-The `.read` method requires one argument, the filename, and allows for additional options. `.read` first opens the CSV file, then reads the contents of it into a new `Array`. Each row of CSV data is translated into an array as well.
+1. What does the variable `record` represent?
+1. For each `record`, what two things does the code do?
+1. What would be a better variable name for `record`?
+1. What data type is each `record`?
+    - What are the keys of each `record`?
+    - What piece of code might have caused this?
+    - What is the data type of the keys?
 
-```
-# planet_data.csv
-1,Mercury,0.055,0.4
-2,Venus,0.815,0.7
-3,Earth,1.0,1.0
-4,Mars,0.107,1.5
-5,Jupiter,318,5.2
-```
+### Another Example, All Together
+
+Observe this code that creates an array of planets that are not Earth:
 
 ```ruby
-require 'csv'
 require 'awesome_print'
-array_of_planet_data = CSV.read("planet_data.csv")
-ap array_of_planet_data
-```
+require 'csv'
 
-## Reading in a file with Headers
-It is also common for a CSV file with a 1st row containing column headers describing the contents. The `read` method accepts an optional parameter allowing you to indicate that the CSV file has a header row. Update the CSV file as follows:
+planets_that_arent_earth = []
 
-```csv
-id,name,mass,distance
-1,Mercury,0.055,0.4
-2,Venus,0.815,0.7
-3,Earth,1.0,1.0
-4,Mars,0.107,1.5
-5,Jupiter,318,5.2
+CSV.read('planets_data.csv', headers: true).each do |planet|
+  if planet["name"] != "Earth"
+    planets_that_arent_earth << planet
+  end
+end
+
+ap planets_that_arent_earth
 ```
 
-Now you can read in the file with:
+1. What does the variable `planets_that_arent_earth` represent? How do we use it?
+1. True or false: We can use conditionals, loops, call methods, read/assign/re-assign variables within this `each` loop.
+
+How can we change this code so that we get an array of planets with a mass smaller than `0.5`?
+
+## Summary
+
+We should observe a pattern that our **source of data, or database** is a system of organizing data that can be separated from our Ruby code.
+
+We can use Ruby code and `CSV.read` to read the CSV file.
+
+We can use a combination of `CSV.read`, the `headers: true` option, and `.each` in order to iterate through this data.
+
+## Bonus Syntax: `CSV.read` and Forcing `CSV::Table` to be an Array
```

Comment:
Will there still be a project and/or activity that requires them to do this? 
If so, I'd prefer we not mark it as bonus to make it clear Instructors should call it out.